### PR TITLE
Fix #639 (Filtering based on NSFW tag)

### LINKED ISSUE #639: feature/Filtering based on NSFW tag
Not Safe For Work images should be tagged with NSFW tag. Then NSFW checkboxes will be added to following pages: Marketplace, Dank Registry - Browse, Leaderboard - Dankest Memes, Memefolio, Related Memes.

NSFW checkbox is unchecked by default.

Home Page will not have NSFW checkbox and will not display images with NSFW tag.

### PR DESCRIPTION
**Pull request by jpmonettas**
At ui level:
 uses check-filters for search boxes
 uses memefactory.ui.components.general/nsfw-switch component for other cases

At server level:
 adds tags-not to meme and auction search schema and corresponding resolvers

### CODE REVIEW

File: src/memefactory/ui/leaderboard/dankest_page.cljs [hunk 1]
```diff
@@ -24,7 +27,8 @@
                        order-by)
             :statuses [:reg-entry.status/whitelisted]
             :order-dir :desc}
-     after (assoc :after after))
+     after       (assoc :after after)
+     nsfw-switch (assoc :tags-not [search/nsfw-tag]))
```
Comment (fbielejec) [inline src/memefactory/ui/leaderboard/dankest_page.cljs]: same here, just declare it as vec

File: src/memefactory/ui/components/general.cljs [hunk 2]
```diff
@@ -17,3 +21,13 @@
                     :href (when route @(subscribe [:district.ui.router.subs/resolve route params query]))}
                    (dissoc props :route :params :query))]
         childs))
+
+(defn nsfw-switch [form-data]
+  (let [id :nsfw-switch]
+    (fn []
```
[the remark this answers is not in the document]
Comment (jpmonettas) [reply src/memefactory/ui/components/general.cljs]: Not needed anymore, it used to contain a ratom but I removed it, so fixed now

File: src/memefactory/ui/components/search.cljs [hunk 3]
```diff
@@ -26,13 +28,19 @@
...
+(def nsfw-tag "nsfw")
+
 (defn search-tools [{:keys [form-data selected-tags-id search-id]}]
 
   (let [search-input-form-data (r/atom {search-id (get @form-data search-id)})
         chip-input-form-data (r/atom {selected-tags-id (get @form-data selected-tags-id)})]
     (fn [{:keys [title sub-title form-data tags selected-tags-id search-id select-options check-filters
-                on-selected-tags-change on-search-change on-select-change option-filters option-filters-id]}]
-     [:div.search-form
+                 on-selected-tags-change on-search-change on-select-change option-filters option-filters-id]}]
```
[the remark this answers is not in the document]
Comment (jpmonettas) [reply src/memefactory/ui/components/search.cljs]: didn't knew we had that convention. Anyway the indentation wasn't right, just reindented the entire form.

File: src/memefactory/ui/components/search.cljs [hunk 4]
```diff
@@ -81,15 +89,17 @@
           :for :tags
           :form-data form-data
           :id [selected-tags-id]}]
-        (if check-filters
-          [:div.check-group (doall
+        (when check-filters
+          [:div.check-group
+           (doall
```
[the remark this answers is not in the document]
Comment (jpmonettas) [reply src/memefactory/ui/components/search.cljs]: this I don't know, it doesn't look wrong to me

File: src/memefactory/ui/dank_registry/browse_page.cljs [hunk 5]
```diff
@@ -16,13 +17,14 @@
...
 
-(defn build-tiles-query [{:keys [:search-term :order-by :search-tags :order-dir :only-cheapest?]} after]
+(defn build-tiles-query [{:keys [:search-term :order-by :search-tags :order-dir :only-cheapest? :nsfw-switch]} after]
   [:search-memes
    (cond-> {:first page-size
             :statuses [:reg-entry.status/whitelisted]}
      (not-empty search-term) (assoc :title search-term)
      (not-empty search-tags) (assoc :tags search-tags)
      after                   (assoc :after after)                    ;; TODO: fix this HACK!
+     nsfw-switch             (assoc :tags-not [search/nsfw-tag])
```
[the remark this answers is not in the document]
Comment (jpmonettas) [reply src/memefactory/ui/dank_registry/browse_page.cljs]: hmm how is that, nsfw-tag is just one tag we don't want to show, but possibly we can add more there in the future, the parameter just get a collection like the other tag parameters

File: src/memefactory/ui/events.cljs [hunk 6]
```diff
@@ -47,3 +47,11 @@
    (let [votes (reader/read-string data-string)]
      {:store (cljs-utils/merge-in store {:votes votes})
       :db (cljs-utils/merge-in db {:votes votes})})))
+
+(re-frame/reg-event-fx
+ ::nsfw-switch
+ [interceptors (re-frame/inject-cofx :store)]
+ (fn [{:keys [db store]} _]
```
[the remark this answers is not in the document]
Comment (jpmonettas) [reply src/memefactory/ui/events.cljs]: didn't knew we had that convention, I like the placeholders there to remember the fn arity, but can remove if it is important

File: src/memefactory/ui/leaderboard/dankest_page.cljs (hunk 1 above)
Comment (jpmonettas) [reply src/memefactory/ui/leaderboard/dankest_page.cljs]: I don't think that should be a vec

File: src/memefactory/ui/dank_registry/browse_page.cljs (hunk 5 above)
[the remark this answers is not in the document]
Comment (fbielejec) [reply src/memefactory/ui/dank_registry/browse_page.cljs]: gotcha

**Pull request merged**